### PR TITLE
refactor: refact type

### PR DESCRIPTION
**Pull request by dl239**
- add `include` dir for user
- move `StringRef` to  top `include` dir
- move `mem_pool`, `Date` and `Timestamp` class to top `include` dir

### CODE REVIEW

**Comment by aceforeverd** [comment]
It's the requirement for C UDF ?

**Comment by dl239** [comment]
> It's the requirement for C UDF ?

yes

**Comment by dl239** [comment]
[C UDF support](https://trello.com/c/53uaUvgk/7-c-udf-support)

File: include/base/type.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef OPENMLDB_INCLUDE_BASE_TYPE_H_
+#define OPENMLDB_INCLUDE_BASE_TYPE_H_
```
Comment (dl239) [inline include/base/type.h:18]: this head guard style has fixed in another pr.

**Comment by zhanghaohit** [comment]
> > It's the requirement for C UDF ?
> 
> yes

what's the motivation of this migration?

**Comment by dl239** [comment]
> > > It's the requirement for C UDF ?
> > 
> > 
> > yes
> 
> what's the motivation of this migration?

the developer will use these class to develop udf. it's better to provide the independent file/dir for users.

**Review by zhanghaohit — APPROVED** [review]
LGTM

**Review by aceforeverd — APPROVED** [review]

**Pull request merged**